### PR TITLE
a new class for realizing Early Stopping

### PR DESCRIPTION
**Pull request by wubowen416**
### Summary

I want to add a new class for realizing Early Stopping. I have already implemented the class in my own code.

It is based on 3 metrics from paper "Early Stopping - but when? Lutz Prechelt, University Karlsrhe". 

1. **Generalization Loss**(GL). GL can be described as a metric that can measure how much current validation loss exceeds the lowest validation loss. Training will stop as soon as GL exceeds a certain threshold. 

2. **Progress Quotient**(PQ).  It considers training strip of length `k`. Progress means how much was the average training error during the strip larger than the minimum training error during the strip. Quotient means use the quotient of GL and Progress. Thus, training will stop as soon as the PQ exceeds a certain threshold. 

3. **UP**. This is simple. Training will stop when the generalization error increased in `s` successive strips.

Combining these metrics, there are 5 modes for doing early stopping. 1, 2, 3, 1+3, 2+3.

I think basically all workers using Keras will be beneficial from this.

Now the EarlyStopping class of Keras can only stop training based on patience and delta, whereas there are advanced and useful early stopping methods.

### Related Issues

This idea comes from others paper, I just realized it. For further info, please refer to the paper.
[early stopping - but when.pdf](https://github.com/keras-team/keras/files/3774689/early.stopping.-.but.when.pdf)

### PR Overview

- [ ] This PR requires new unit tests [n] (make sure tests are included)
- [x] This PR requires to update the documentation [y] (make sure the docs are up-to-date)
- [ ] This PR is backwards compatible [y/n]
- [ ] This PR changes the current API [n] (all API changes need to be approved by fchollet)

### CODE REVIEW

**Comment by gabrieldemarmiesse** [comment]
Thank you for this cool tool! Could you add it to tensorflow/addons? This is the repo made for new tools before we're sure to include them in the core library.
